### ISSUE TITLE
Temp buffer (floatwindow) with go type cause high CPU usage of gopls

### ISSUE THREAD

**Issue by ray-x** [neovim/nvim-lspconfig#943]
The issue found is when I use Vista + gopls. Vista can preview tag/function in floatwindows. 
When I open preview, Vista will show a pretty floating window with highlight with go syntax (by setting file type to go)

If my gopls rootpath is 
```lua
{
root_dir = function(fname)
      return util.root_pattern("go.mod", ".git")(fname) or util.path.dirname(fname)
    end
}
```
`util.path.dirname(fname)` part give me a problem.
It returns an empty string (not nil)
That empty string will be pass to rootpath and as there is no separator. In the end, it becomes "/", which is a root dir in mac.

That will ask gopls to search all files in "/" and slow down gopls.

I think a fix should address this. I will submit a PR.

**Comment by mjlbach** [comment]
Shouldn’t this be the fallback anyways?

________________________________
From: rayx ***@***.***>
Sent: Thursday, May 27, 2021 3:15:23 PM
To: neovim/nvim-lspconfig ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [neovim/nvim-lspconfig] Temp buffer (floatwindow) with go type cause high CPU usage of gopls (#943)


The issue found is when I use Vista + gopls. Vista can preview tag/function in floatwindows.
When I open preview, Vista will show a pretty floating window with highlight with go syntax (by setting file type to go)

If my gopls rootpath is

{
root_dir = function(fname)
      return util.root_pattern("go.mod", ".git")(fname) or util.path.dirname(fname)
    end
}

util.path.dirname(fname) part give me a problem.
It returns an empty string (not nil)
That empty string will be pass to rootpath and as there is no separator. In the end, it becomes "/", which is a root dir in mac.

That will ask gopls to search all files in "/" and slow down gopls.

I think a fix should address this. I will submit a PR.

**Comment by ray-x** [comment]
I think "/" is the fallback...

**Comment by ray-x** [comment]
go.mod and .git is not essential to a go project folder. 
e.g a folder only contain  hello_world.go, I can still use `go run hello_world.go` to run the code

It is true that by default, the gopls setting works perfectly.

**Comment by mjlbach** [comment]
Yeah but in that case the current fallback *should* work, returning early isn't the correct fix...

**Comment by ray-x** [comment]
for function dirname(filename)
if filename is nil or empty string. both should return directly.
Further. I think we can only return "/" (root dir) only if it is "/".   Use "/" as the default return seems may cause some unexpected behaviour.

**Comment by mjlbach** [comment]
Closed by #944
